documentation/proposals/Proposal - Silk.NET 3.X Generic Maths.md

```diff
@@ -0,0 +1,884 @@
+    - ILogarithmicFunctions
+    - IExponentialFunctions
+    - IPowerFunctions
+    - IRootFunctions
+    (The reasoning for this is that the required interface types are more narrow than INumberBase)
+- Static helper methods, where the result is expected to be bounded to 0-1 or the type of T has to be scoped to a type narrower than the one used by the primitive, have two variants:
+    - Method&lt;T>: T is scoped to a narrower type (e.g IRootFunctions&lt;T>, IFloatingPoint&lt;T>)
+    - Method&lt;T, TReturn>: no scope for T, but TReturn is scoped to a narrower type
+
+# Proposed API
+- IVector&lt;TVector, T>
+- IVectorAlso&lt;TVector, T> - Covers interfaces that can't be implemented by IVector because of compiler restrictions
+- Vector2D&lt;T>
+- Vector3D&lt;T>
+- Vector4D&lt;T>
+- Quaternion&lt;T>
+- Matrix3x2&lt;T>
+- Matrix4x4&lt;T>
+- Plane&lt;T>
+
+For all `VectorND<T>`:
+```cs
+public readonly partial struct Vector2D<T> : IVector<Vector2D<T>, T>, IVectorAlso<Vector2D<T>, T>, IEquatable<Vector2>, IVec2
+    where T : INumberBase<T>
+{
+    /// <summary>The X component of the vector.</summary>
+    public readonly T X;
+
+    /// <summary>The Y component of the vector.</summary>
+    public readonly T Y;
+
+    internal const int Count = 2;
+
+    /// <summary>Creates a new <see cref="Vector2D{T}" /> object whose two elements have the same value.</summary>
+    /// <param name="value">The value to assign to all two elements.</param>
+    public Vector2D(T value);
+
+    /// <summary>Creates a vector whose elements have the specified values.</summary>
+    /// <param name="x">The value to assign to the <see cref="X" /> field.</param>
+    /// <param name="y">The value to assign to the <see cref="Y" /> field.</param>
+    public Vector2D(T x, T y);
+
+    /// <summary>Constructs a vector from the given <see cref="ReadOnlySpan{T}" />. The span must contain at least 4 elements.</summary>
+    /// <param name="values">The span of elements to assign to the vector.</param>
+    public Vector2D(ReadOnlySpan<T> values);
+
+    /// <summary>Gets a vector whose 4 elements are equal to zero.</summary>
+    /// <value>A vector whose two elements are equal to zero (that is, it returns the vector <c>(0,0)</c>.</value>
+    public static Vector2D<T> Zero => new(T.Zero);
+
+    /// <summary>Gets a vector whose 4 elements are equal to one.</summary>
+    /// <value>Returns <see cref="Vector2D{T}" />.</value>
+    /// <remarks>A vector whose two elements are equal to one (that is, it returns the vector <c>(1,1)</c>.</remarks>
+    public static Vector2D<T> One => new(T.One);
+
+    /// <summary>Gets the vector (1,0)).</summary>
+    /// <value>The vector <c>(1,0)</c>.</value>
+    public static Vector2D<T> UnitX => new(T.One, T.Zero);
+
+    /// <summary>Gets the vector (0,1)).</summary>
+    /// <value>The vector <c>(0,1)</c>.</value>
+    public static Vector2D<T> UnitY => new(T.Zero, T.One);
+
+    public ReadOnlySpan<T> Components { get; }
```

Comment:
Why not `AsSpan()` instead, which matches the convention we use in the BCL.

Reply:
Not sure what my logic was with this here. Will change.